common/engine_conf/message_helper.hh

```diff
@@ -0,0 +1,230 @@
+class anomalydetection_helper;
+class severity_helper;
+class tag_helper;
+class state_helper;
+
+using ::google::protobuf::Message;
+
+bool fill_pair_string_group(PairStringSet* grp, const std::string_view& value);
+bool fill_pair_string_group(PairStringSet* grp,
+                            const std::string_view& key,
+                            const std::string_view& value);
+void fill_string_group(StringList* grp, const std::string_view& value);
+void fill_string_group(StringSet* grp, const std::string_view& value);
+bool fill_host_notification_options(uint32_t* options,
+                                    const std::string_view& value);
+bool fill_service_notification_options(uint32_t* options,
+                                       const std::string_view& value);
+
+/**
+ * @brief The base message helper used by every helpers. It defines the common
+ * methods.
+ *
+ */
+class message_helper {
+ public:
+  enum object_type {
+    command = 0,
+    connector = 1,
+    contact = 2,
+    contactgroup = 3,
+    host = 4,
+    hostdependency = 5,
+    hostescalation = 6,
+    hostgroup = 8,
+    service = 9,
+    servicedependency = 10,
+    serviceescalation = 11,
+    servicegroup = 13,
+    timeperiod = 14,
+    anomalydetection = 15,
+    severity = 16,
+    tag = 17,
+    state = 18,
+  };
+
+ private:
+  const object_type _otype;
+  Message* _obj;
+  const absl::flat_hash_map<std::string, std::string> _correspondence;
+  std::vector<bool> _modified_field;
+  bool _resolved = false;
+
+ public:
+  message_helper(object_type otype,
+                 Message* obj,
+                 absl::flat_hash_map<std::string, std::string>&& correspondence,
+                 size_t field_size)
+      : _otype(otype),
+        _obj(obj),
+        _correspondence{
+            std::forward<absl::flat_hash_map<std::string, std::string>>(
+                correspondence)},
+        _modified_field(field_size, false) {}
+  message_helper(const message_helper& other);
+  message_helper() = delete;
+  message_helper& operator=(const message_helper&) = delete;
+  virtual ~message_helper() noexcept = default;
+  const absl::flat_hash_map<std::string, std::string>& correspondence() const {
+    return _correspondence;
+  }
+  object_type otype() const { return _otype; }
+  Message* mut_obj() { return _obj; }
+  const Message* obj() const { return _obj; }
+  void set_obj(Message* obj) { _obj = obj; }
+  void set_changed(int num) { _modified_field[num] = true; }
+  bool changed(int num) const { return _modified_field[num]; }
+  bool resolved() const { return _resolved; }
+  void resolve() { _resolved = true; }
+
+  /**
+   * @brief For several keys, the parser of objects has some particular
+   *        behaviors. These behaviors are handled here.
+   * @param key The key to parse.
+   * @param value The value corresponding to the key
+   *
+   * @return True on success.
+   */
+  virtual bool hook(std::string_view key [[maybe_unused]],
```

Comment:
whhy not a reference for key
```suggestion
  virtual bool hook(const std::string_view & key [[maybe_unused]],
```

Reply:
Since there is a call to a validate_key() method inside this method, it is faster to copy the key as parameter, because we just have to modify it after instead of allocating another one.